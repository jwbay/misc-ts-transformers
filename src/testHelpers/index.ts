import { stripIndent } from 'common-tags'
import { SourceMapConsumer } from 'source-map'
import {
	JsxEmit,
	NewLineKind,
	ScriptTarget,
	transpileModule,
} from 'typescript'
import { FileTransformer } from '../helpers'

export function createAsserter(transformer: FileTransformer) {
	return function testCase(testName: string, content: string, only = false) {
		const fn = only ? test.only : test
		fn(testName, () => {
			expect(transform(content, transformer)).toMatchSnapshot()
		})
	}
}

function transform(source: string, transformer: FileTransformer) {
	const { outputText, sourceMapText, diagnostics } = transpileModule(source, {
		compilerOptions: {
			jsx: JsxEmit.React,
			newLine: NewLineKind.LineFeed,
			sourceMap: true,
			target: ScriptTarget.ES5,
		},
		fileName: 'module.tsx',
		reportDiagnostics: true,
		transformers: {
			after: [transformer],
		},
	})

	if (diagnostics && diagnostics.length > 0) {
		diagnostics.forEach(error => {
			console.error(stripIndent`
				${error.messageText}
				${source.substr(Math.min(0, error.start - 5), error.length + 10)}
			`)
		})

		throw new Error('Syntax error')
	}

	return {
		code: createSnapshotFriendlySourceCode(outputText),
		sourceMap: createSnapshotFriendlySourceMap(sourceMapText as string),
	}
}

function createSnapshotFriendlySourceCode(content: string) {
	return '\n' + content
		.replace('//# sourceMappingURL=module.js.map', '')
		.replace(/"/g, `'`)
}

function createSnapshotFriendlySourceMap(sourceMapText: string) {
	const smc = new SourceMapConsumer(sourceMapText)
	const mappings: { [line: number]: string[] } = {}
	smc.eachMapping(({ generatedColumn, generatedLine, originalColumn, originalLine }) => {
		(mappings[generatedLine] || (mappings[generatedLine] = [])).push(
			`[${generatedLine},${generatedColumn} => ${originalLine},${originalColumn}]`,
		)
	})

	const result: { [line: string]: string } = {}
	Object.keys(mappings).forEach(line => {
		result['line ' + line] = mappings[line].join(', ')
	})
	return result
}
