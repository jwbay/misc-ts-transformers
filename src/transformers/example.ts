import {
	createIdentifier,
	Identifier,
	setTextRange,
	SyntaxKind,
} from 'typescript'
import { FileTransformer, transformKinds } from '../helpers'

export const replaceIdentifiersNamedOldNameWithName: FileTransformer = transformKinds({
	[SyntaxKind.Identifier]: (node: Identifier) => {
		if (node.text === 'oldName') {
			return setTextRange(createIdentifier('name'), node)
		}
		return node
	},
})
