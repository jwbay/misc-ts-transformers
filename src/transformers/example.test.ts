import { stripIndent } from 'common-tags'
import { createAsserter } from '../testHelpers'
import { replaceIdentifiersNamedOldNameWithName } from './example'

const testCase = createAsserter(replaceIdentifiersNamedOldNameWithName)

testCase('works', stripIndent`
	const oldName = 42;
	function asdf(oldName) {
		console.log(oldName);
	}
	asdf(oldName);
`)

testCase('works too', stripIndent`
	class oldName {
		private oldName = 42;
	}
`)
