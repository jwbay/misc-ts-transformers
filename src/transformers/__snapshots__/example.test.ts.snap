// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "code": "
var name = 42;
function asdf(name) {
    console.log(name);
}
asdf(name);
",
  "sourceMap": Object {
    "line 1": "[1,0 => 1,0], [1,4 => 1,6], [1,8 => 1,13], [1,11 => 1,16], [1,13 => 1,18], [1,14 => 1,19]",
    "line 2": "[2,0 => 2,0], [2,14 => 2,14], [2,18 => 2,21]",
    "line 3": "[3,4 => 3,1], [3,11 => 3,8], [3,12 => 3,9], [3,15 => 3,12], [3,16 => 3,13], [3,20 => 3,20], [3,21 => 3,21], [3,22 => 3,22]",
    "line 4": "[4,0 => 4,0], [4,1 => 4,1]",
    "line 5": "[5,0 => 5,0], [5,4 => 5,4], [5,5 => 5,5], [5,9 => 5,12], [5,10 => 5,13], [5,11 => 5,14]",
  },
}
`;

exports[`works too 1`] = `
Object {
  "code": "
var name = (function () {
    function name() {
        this.name = 42;
    }
    return name;
}());
",
  "sourceMap": Object {
    "line 1": "[1,0 => 1,0], [1,4 => 1,6], [1,8 => 1,13]",
    "line 2": "[2,4 => 1,0]",
    "line 3": "[3,8 => 2,9], [3,17 => 2,16], [3,20 => 2,19], [3,22 => 2,21], [3,23 => 2,22]",
    "line 4": "[4,4 => 3,0], [4,5 => 3,1]",
    "line 5": "[5,4 => 3,0], [5,11 => 1,6], [5,15 => 3,1]",
    "line 6": "[6,0 => 3,0], [6,1 => 1,0], [6,1 => 3,1], [6,5 => 3,1]",
  },
}
`;
